llama_index/llms/llm_prompt_mixin.py

```diff
@@ -0,0 +1,254 @@
+from collections import ChainMap
+from typing import (
+    Any,
+    Callable,
+    List,
+    Optional,
+)
+
+from llama_index.bridge.pydantic import BaseModel, Field
+from llama_index.callbacks import CBEventType, EventPayload
+from llama_index.llms.base import LLM
+from llama_index.llms.types import (
+    ChatMessage,
+    ChatResponseAsyncGen,
+    ChatResponseGen,
+    CompletionResponseAsyncGen,
+    CompletionResponseGen,
+    MessageRole,
+)
+from llama_index.prompts import BasePromptTemplate, PromptTemplate
+from llama_index.types import PydanticProgramMode, TokenAsyncGen, TokenGen
+
+
+def stream_completion_response_to_tokens(
+    completion_response_gen: CompletionResponseGen,
+) -> TokenGen:
+    """Convert a stream completion response to a stream of tokens."""
+
+    def gen() -> TokenGen:
+        for response in completion_response_gen:
+            yield response.delta or ""
+
+    return gen()
+
+
+def stream_chat_response_to_tokens(
+    chat_response_gen: ChatResponseGen,
+) -> TokenGen:
+    """Convert a stream completion response to a stream of tokens."""
+
+    def gen() -> TokenGen:
+        for response in chat_response_gen:
+            yield response.delta or ""
+
+    return gen()
+
+
+async def astream_completion_response_to_tokens(
+    completion_response_gen: CompletionResponseAsyncGen,
+) -> TokenAsyncGen:
+    """Convert a stream completion response to a stream of tokens."""
+
+    async def gen() -> TokenAsyncGen:
+        async for response in completion_response_gen:
+            yield response.delta or ""
+
+    return gen()
+
+
+async def astream_chat_response_to_tokens(
+    chat_response_gen: ChatResponseAsyncGen,
+) -> TokenAsyncGen:
+    """Convert a stream completion response to a stream of tokens."""
+
+    async def gen() -> TokenAsyncGen:
+        async for response in chat_response_gen:
+            yield response.delta or ""
+
+    return gen()
+
+
+class LLMPromptMixin(LLM):
```

Comment:
I don't think this is really a Mixin.

Reply:
A mixin pattern is more like

```
class LLMPromptMixin:
    ...


class LLMImpl(LLM, LLMPromptMixin): 
    ...
```

where LLMImpl inherits from the base class LLM, and then inject attributes / behaviors from LLMPromptMixin

Reply:
what you have here is functionally equivalent to just adding these attributes / behaviors directly to the LLM base class?

Reply:
An actual mixin is less helpful here I think... because then we can't rely on typing functions with `LLM` and expect them to have the new methods